### PR TITLE
Persimmon fa2 attention4d

### PR DESCRIPTION
**Pull request by jeromeku**
# What does this PR do?
Adds Flash Attention 2 for Persimmon per  #26350 
Adds 2d->4d attention mask per #26792

## Who can review?
@younesbelkada

### CODE REVIEW

**Comment by jeromeku** [comment]
@younesbelkada 

lmk if my implementation of 4d attention mask (#26792) + FA2 needs tweaking.

Regarding previous [comment](https://github.com/huggingface/transformers/issues/26350?utm_source=tldrai#issuecomment-1745613947), I'd like to understand HF's current strategy for integrating third-party / OSS libraries and components.  Given the rapid pace of innovation in this space, want to ensure that `transformers` and its sister libraries remain best-in-class wrt usability and performance!

**Comment by younesbelkada** [comment]
Thanks very much for your great contrib @jeromeku ! Sorry for the delay responding on the PR, I will have an extensive look at the PR and your questions by beginning of next week (from 30rd october) 🙏

**Comment by pszemraj** [comment]
Hi! great work and I don't mean to butt in here, but in case it helps take this home:

I was trying to get this working and ran into some issues with the latest (4.36.dev0) version of transformers after cloning this pr and rebasing on main. I had to this because of the llama2 tokenizer/optimum import issue that I get using verbatim the transformers version **on this PR as-is**.

After scouring gh, I came across a fine-tuning repo for Fuyu, and the same author has a working version of FA2 for persimmon (I was able to train persimmon on it with FA2): 
- https://github.com/phillip-kravtsov/transformers/tree/floating-updates

```sh
pip install git+https://github.com/phillip-kravtsov/transformers.git@floating-updates
```

This is experimental and the work of one dude, so for the record the working SHA is: `b8000bd8d619cbbedcb806b67faa68c2300b4bd0`


hope this helps!

**Comment by jeromeku** [comment]
@younesbelkada 

Let me know how I can improve the PR.  Also, would appreciate thoughts on previous [query](https://github.com/huggingface/transformers/issues/26350#issuecomment-1745613947) when you get a chance.

Thanks!

File: src/transformers/models/persimmon/modeling_persimmon.py [hunk 1]
```diff
@@ -27,17 +27,169 @@
...
+    from flash_attn.bert_padding import index_first_axis, pad_input, unpad_input  # noqa
+
+
 logger = logging.get_logger(__name__)
 
 _CONFIG_FOR_DOC = "PersimmonConfig"
 
 
+# Copied from transformers.models.llama.modeling_llama.AttnMaskConverter
+class AttnMaskConverter:
```
Comment (younesbelkada) [inline src/transformers/models/persimmon/modeling_persimmon.py:57]: instead you can import it from `modeling_attn_mask_utils`:
```python
from ...modeling_attn_mask_utils import AttentionMaskConverter, _prepare_4d_causal_attention_mask
```

**Review by younesbelkada — COMMENTED** [review]
Great work! Thank you @jeromeku - I left one comment
Can you also add few lines in the documentation? You can do something similar than : https://github.com/huggingface/transformers/pull/27400/

File: src/transformers/models/persimmon/modeling_persimmon.py (hunk 1 above)
Comment (jeromeku) [reply src/transformers/models/persimmon/modeling_persimmon.py:57]: @younesbelkada 

Made the aforementioned changes but am getting odd errors with `test_flash_attn_2_generate_padding_right` test (all other tests pass).  

For example, here are the outputs with and without FA2:
```python
out=tensor([[38, 17, 72, 39, 27, 98, 50, 53],
        [88, 83, 53, 72, 98, 43, 42, 90],
        [44, 38, 53, 70, 57, 77, 58, 87],
        [61, 57, 26, 71, 68,  5, 76, 60],
        [29,  8, 61, 62, 66, 12, 52, 63],
        [83, 71, 43, 61, 22, 83, 89, 89],
        [50, 38, 11, 22, 42, 13, 65, 14],
        [71, 71,  0,  9,  3, 37, 10, 84],
        [81, 19, 56, 62, 67, 18, 58, 87],
        [39,  9, 49, 48, 22, 13, 72, 20],
        [26, 96,  7, 26, 54, 46, 32,  6],
        [67,  9, 87, 93, 42,  8, 65, 14],
        [25, 83, 70, 30, 32, 92, 25, 64]], device='cuda:0')
out_fa=tensor([[38, 17, 72, 39, 27, 98, 50, 95],
        [88, 83, 53, 72, 98, 43, 42, 30],
        [44, 38, 53, 70, 57, 77, 58, 87],
        [61, 57, 26, 71, 68,  5, 76, 60],
        [29,  8, 61, 62, 66, 12, 52, 28],
        [83, 71, 43, 61, 22, 83, 89, 89],
        [50, 38, 11, 22, 42, 13, 65, 10],
        [71, 71,  0,  9,  3, 37, 10, 84],
        [81, 19, 56, 62, 67, 18, 58, 87],
        [39,  9, 49, 48, 22, 13, 72, 10],
        [26, 96,  7, 26, 54, 46, 32, 94],
        [67,  9, 87, 93, 42,  8, 65, 10],
        [25, 83, 70, 30, 32, 92, 25, 64]], device='cuda:0')
```
Seems like there is some unaccounted for randomness as some sequences match while are others do not.

This is odd given that the `flash_attn_2_inference_padding_{left,right}` tests passes as does `flash_attn_2_generate_left_padding` test.

Any ideas what could be causing this?

**Comment by jeromeku** [comment]
@younesbelkada 

Trying to get to the bottom of the issue:
- Created a fresh `venv` and cloned the `transformers` repo as of `11/20/2023` and did a `pip install -e .[dev]`
- Installed `flash_attn`:
```
Name: flash-attn
Version: 2.3.4
Summary: Flash Attention: Fast and Memory-Efficient Exact Attention
Home-page: https://github.com/Dao-AILab/flash-attention
Author: Tri Dao
Author-email: trid@cs.stanford.edu
License: 
Location: /notebooks/virtualenvs/transformers-test/lib/python3.9/site-packages
Requires: einops, ninja, packaging, torch
Required-by: 
Metadata-Version: 2.1
Installer: pip
Classifiers:
  Programming Language :: Python :: 3
  License :: OSI Approved :: BSD License
  Operating System :: Unix
Entry-points:
Project-URLs:
```
- Ran `attn_2` tests for the following models (only showing failures):

`Whisper`
```
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelTest::test_flash_attn_2_inference - AssertionError: assert False
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperStandaloneDecoderModelTest::test_flash_attn_2_generate_padding_right - AssertionError: False is not true
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperStandaloneDecoderModelTest::test_flash_attn_2_inference - AssertionError: assert False
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperStandaloneDecoderModelTest::test_flash_attn_2_inference_padding_right - AssertionError: assert False
==============
```
 `Mistral`
```
FAILED tests/models/mistral/test_modeling_mistral.py::MistralModelTest::test_flash_attn_2_generate_padding_right - AssertionError: ValueError not raised
FAILED tests/models/mistral/test_modeling_mistral.py::MistralModelTest::test_flash_attn_2_inference_padding_right - AssertionError: ValueError not raised
```
`Bark`
```
FAILED tests/models/bark/test_modeling_bark.py::BarkSemanticModelTest::test_flash_attn_2_fp32_ln - RuntimeError: FlashAttention only support fp16 and bf16 data type
FAILED tests/models/bark/test_modeling_bark.py::BarkSemanticModelTest::test_flash_attn_2_from_config - ValueError: Unrecognized configuration class <class 'transformers.models.bark.configuration_bark.BarkSemanticConfig'> for this kind of AutoModel: ...
FAILED tests/models/bark/test_modeling_bark.py::BarkCoarseModelTest::test_flash_attn_2_fp32_ln - RuntimeError: FlashAttention only support fp16 and bf16 data type
FAILED tests/models/bark/test_modeling_bark.py::BarkCoarseModelTest::test_flash_attn_2_from_config - ValueError: Unrecognized configuration class <class 'transformers.models.bark.configuration_bark.BarkCoarseConfig'> for this kind of AutoModel: Au...
```
`GPTNeo`
```
FAILED tests/models/gpt_neo/test_modeling_gpt_neo.py::GPTNeoModelTest::test_flash_attn_2_generate_padding_right - AssertionError: False is not true
```
`Llama`, `Distillbert`, `GPT BigCode` tests all pass.  
`Falcon` I get `cuda device-side error`, which might be due to the fact that I'm running on an `A6000` (48Gb memory) which might not be sufficient. 

Thoughts?

**Comment by susnato** [comment]
Hi @jeromeku, the `test_flash_attn_2_generate_padding_right` for `GptNeo` is quite flaky, most of the time it passes but sometimes it fails.

**Pull request closed without merging**